Tests/PositionlessTests/RotateTest.swift

```
@@ -0,0 +1,57 @@
+import Testing
+
+@testable import Positionless
+
+@Test func whenFirstHalfIsEmpty() {
+  var array = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+  array.withMutablePartition(into: 2) { partition in
+    partition.rotate()
+    #expect(partition.partitionCount == 2)
+    #expect(partition[part: 0].equals([1, 2, 3, 4, 5, 6, 7, 8, 9]))
+    #expect(partition[part: 1].equals([]))
+  }
+}
+
+@Test func whenSecondHalfIsEmpty() {
+  var array = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+  array.withMutablePartition(into: 2) { partition in
+    partition.grow(part: 0, by: 9)
+    partition.rotate()
+    #expect(partition.partitionCount == 2)
+    #expect(partition[part: 0].equals([]))
```

I haven't tried this yet, but I expect that `==` should work better than `.equals()`, as the `#expect` macro can print both sides of the equality when failure happens instead of just printing expected `True` but was `False`. It can help with diagnosing problems.
(Not very important, just something to try next time.)

---

I actually wanted to have the same. However, I don't know if I can implement Equatable protocol for Collection in a generic way. Thus I needed to have equals method. If that is possible, we should definitely use == then.

---

Sorry for the late response, I now realized I had github notifications turned off.

Oh I see, that's interesting. Especially, if your collection is like a forward iterator and once you read it it's gone, equality checking may not be always possible in a non-mutating way. I would then not require Collection to conform to `Equatable` but you can still conform individual concrete types. 

Also, you can create blanket conformances if that's applicable: you can express implications with Swift generics such as if the element type or underlying collection type is Equatable, make the wrapper `SegmentedCollection` also equatable, relying on calling the Equatable implementations of the parts that you are now sure also equatable.

I couldn't yet review the details of your implementation, this is just an example showcasing the technique:
```swift
protocol Container {
    associatedtype ItemID
    associatedtype Content

    var id: ItemID { get }
    var content: Content { get }
}

// make the container equatable when both ItemID and Content are equatable
extension Container where ItemID: Equatable, Content: Equatable {
    static func == (lhs: Self, rhs: Self) -> Bool {
        return lhs.id == rhs.id && lhs.content == rhs.content
    }
}
```

---

Hi! Sorry, I was busy with some work presentations for a week.

> if your collection is like a forward iterator and once you read it it's gone

That's not the case with forward iterator, I think that's the case with input iterators as forward iterators are multi-pass in nature. So, we don't need to mutate collections for equality check.

I actually tried the overloading `==` operator for Collections in generic way. And it works so nicely!!! I have put my changes in [this branch](https://github.com/hylo-lang/positionless-swift/tree/equality_for_collections).

Despite this is good for swift-testing framework, I have few concerns with this `==` overload. Let me discuss this with you:
1. Coming from [EOP](https://www.elementsofprogramming.com/eop.pdf), Alex and Paul write: `Equality is a procedure that takes two objects of the same type and returns true if and only if the object states are equal.`. This indicates `==` should be on same type. However, they also gave a project `Extend the notions of equality, assignment, and copy construction to objects of distinct types`. TBH I don't know the actual solution to this project, maybe this is the usecase they were indicating.
2. When defining `==` generically for Collections, we are essentially saying that "with equality check, elements and their ordering matters, not the shape of structure that contains the elements". TBH I am little bit uncomfortable to this thing. For that matter, I think even my naming `equals` is not correct and should be `elements_equal`.

---

Overloading is just a fancy way of giving the same name to different operations. So we can pretend that one of the `==` meets Alex's definition and the others happen to share the same name. Now, if we consider the code post-elaboration (i.e., once name resolution has concluded) there is no longer any ambiguity as to what a particular use of a resolved `==` means.

I think requiring that a particular _unqualified_ name implements one specific operation is too strict of a constraint. The fully qualified name of `==` under Alex's definition is `Equatable.==` and this one does have the type `T -> T -> Bool`.

---

Actually we can demonstrate that the type of an object (unless the notional value is defined otherwise) is not a salient property of that object and therefore it does not matter whether the two sides of `==` have the same type or not. It suffices to widen the types on both sides to a common type, which always exists in a join-semillattice. Now we can define `==` in terms of that join and we're done.

In less fancy academic mumbo jumbo:
```
func ==(_ x: Any, _ y: Any) -> Bool {
  // probably very slow impl., but the type system doesn't care
}

let x: String = "a"
let y: Int = 1
let b: Bool = x == y
```
By subsumption we can widen `String` to `Any` and `Bool` to `Any` and we'll apply `==` with type `Any -> Any -> Bool`.

---

@kyouko-taiga It's not nearly so cut-and-dried as that.  There may be no common `Equatable` type that's the result of widening and you may have no way to _get_ to such a type.  That's true even if both sides are numeric unless you carefully maintain a representation of the join semilattice and to do that you have to know what all the other types are.  Consider complex numbers using floating point and integer rationals, for example.  And in your example, the problem isn't the performance of the implementation of your equality over `Any`s, but its implementation.  
[You surely know this but just to get it off the table: the one special property that equality has, that makes it possible to implement `==` for `any Equatable` is that you can always say two objects of distinct types are unequal, but that defeats the purpose of the exercise].
Lastly, I don't think “giving the same name to different operations” should be taken lightly, or maybe shouldn't even happen at all, except by accident.  We shouldn't use `==` to describe arbitrary predicates for the same reason it is wrong to call a random shuffle `sort`.  A function called `==` should represent some notion of equality of both its arguments.

@RishabhRD regarding point #2, IMO the key issue to consider is “what constitutes the notional value of the collection?” If a collection is notionally just its elements and their ordering, like arrays and linked lists, for example, internal structure truly doesn't matter to their values.  You can see that by considering a generic function that operates on either an array or a linked list and seeing that it would be valid and meet its specification regardless of intermal structure.  The only possible difference in value would be that an array has a `capacity`-before-reallocation, but we say that is *incidental* to its value.  We can justify saying that because approximately no code that gets an array parameter ever cares about its `capacity`.  On the other hand, if a collection's value is notionally more/different than its elements and their ordering, maybe it is playing too many roles and shouldn't itself be a collection.  For an example of that, in Hylo (but not Swift) we say that hashed sets and dictionaries are not themselves collections, but they have an incidental properties that present their elements as collections.

---

@dabrahams got your point. Notion of Collection is for sure its elements and their ordering. Hash set or dictionary was my point of concern, but collection as incidental property for them just solves that.

I am then merging branch that provides `==` operator for any 2 collections and uses that everywhere.